Implement constructor, initial bootstrap, and tests, for `PropagationNetwork`.

lemonholic opened the pull request:
## Type of this PR
- New feature (non-breaking change which adds functionality)
- Change in tests


## Describe your changes
### Interface implementations
Among the following interfaces, `async fn new`(the constructor) and `async fn create_recv_queue` are implemented. The rest will be available after broadcasting logic is implemented.
```rust
#[async_trait]
pub trait AuthorizedNetwork {
    /// Joins the network with an authorized identity.
    async fn new([OMITTED]) -> Result<Self, String> where Self: Sized;
    /// Broadcasts a message to the network, after signed by the key given to this instance.
    async fn broadcast(&self, message: &[u8]) -> Result<BroadcastToken, String>;
    /// Stops a currently broadcasting message.
    async fn stop_broadcast(&self, token: BroadcastToken) -> Result<(), String>;
    /// Gets the current status of a broadcasting message.
    async fn get_broadcast_status(&self, token: BroadcastToken) -> Result<BroadcastStatus, String>;
    /// Creates a receiver for every message broadcasted to the network, except the one sent by this instance.
    async fn create_recv_queue(&self) -> Result<broadcast::Receiver<Vec<u8>>, ()>;
    /// Provides the estimated list of live nodes that are eligible and identified by their public keys.
    async fn get_live_list(&self) -> Result<Vec<PublicKey>, ()>;
}
```

### Other implemented features
- Provide configurations for some parameters used to construct and maintain the network.
- Implement initial bootstrap, which is run once and while the `PropagationNetwork` is being constructed.
- Create background task that listens on network events (such as incoming connection requests) and handles them to provide public interfaces.
- Message passing between `PropagationNetwork` and the background task.

### Unit tests
The following unit tests for initial bootstrap are added.
```rust
/// Test if every node fills its routing table with the addresses of all the other nodes
/// in a tiny network when they join the network sequentially.
/// (network_size = 5 < [`libp2p::kad::K_VALUE`] = 20)
async fn discovery_with_tiny_network_sequential();

/// Test if every node fills its routing table with the addresses of all the other nodes
/// in a small network when they join the network sequentially.
/// (network_size = [`libp2p::kad::K_VALUE`] = 20)
async fn discovery_with_small_network_sequential();

/// Test if every node fills its routing table with the addresses of all the other nodes
/// in a tiny network when they join the network at the same time.
/// (network_size = 5 < [`libp2p::kad::K_VALUE`] = 20)
async fn discovery_with_tiny_network_concurrent();

/// Test if every node fills its routing table with the addresses of all the other nodes
/// in a small network when they join the network at the same time.
/// (network_size = [`libp2p::kad::K_VALUE`] = 20)
async fn discovery_with_small_network_concurrent();
```

## Remaining works
*The followings are ordered by priority but can be changed.*
- Provide error types for this module(`simperby-network`)
- Add tests for broadcasting
- Implement public interfaces related to broadcasting

posgnu commented:
This PR is good in terms of that it is self-contained, though, it is too big. 
https://gist.github.com/mikepea/863f63d6e37281e329f8#aim-for-one-succinct-commit

junha1 reviewed (commented):
1. Please use `does ~` instead of `do ~` for doc-comments.
2. What is `K_VALUE`?

lemonholic commented:
Rebased on latest commit of main

lemonholic commented on network/src/propagation_network/mod.rs:
```diff
@@ -45,65 +77,501 @@ impl AuthorizedNetwork for PropagationNetwork {
+    /// Assigned ports can be used by dial attempts or other processes.
+    struct PortDispenser {
+        // 10000 ~ 65535
+        available_ports: Mutex<HashSet<u16>>,
```
Reduced the range to 55000..57000, slightly larger than what you've suggested, for further tests.

lemonholic commented:
> 1. Please use `does ~` instead of `do ~` for doc-comments.
> 2. What is `K_VALUE`?

1. Fixed.
2. It is the maximum number of connections(except for one-time connections) or the maximum number of nodes in contact. I replaced it in the comment with descriptive text.

lemonholic commented on network/src/propagation_network/mod.rs:
```diff
@@ -45,65 +77,501 @@ impl AuthorizedNetwork for PropagationNetwork {
+}
+
+impl PropagationNetwork {
+    #[allow(dead_code)]
```
moved it

lemonholic commented on network/src/propagation_network/config.rs:
```diff
@@ -0,0 +1,83 @@
+    pub(crate) peer_discovery_interval: Duration,
+    /// Capacity for the message queue that passes messages from other nodes
+    /// to its simperby node.
+    pub(crate) message_queue_capacity: usize,
```
Applied it to all structs.

lemonholic commented on network/src/propagation_network/behaviour.rs:
```diff
@@ -1,30 +1,64 @@
+            floodsub: Floodsub::new(local_peer_id),
+        }
+    }
+    // Todo: Add constructor taking a configuration as an argument.
```
Just removed the comment because it is already achieved.

lemonholic commented on network/src/propagation_network/mod.rs:
```diff
@@ -45,65 +77,500 @@ impl AuthorizedNetwork for PropagationNetwork {
+        if let Ok(keypair_inner) = ed25519::Keypair::decode(&mut keypair_bytes) {
+            Ok(Keypair::Ed25519(keypair_inner))
+        } else {
+            Err("Invalid public/private keypair was given.".to_string())
```
Applied it to all err messages and panic messages.

lemonholic commented on network/src/propagation_network/config.rs:
```diff
@@ -0,0 +1,83 @@
+    /// Returns a default.
+    ///
+    /// To customize configurations, call `default` and chain the functions with the name of `with_<fieldname>`.
+    pub fn default() -> Self {
```
Moved it into `impl Default` block.
